node/network/bridge/src/lib.rs

```diff
@@ -0,0 +1,999 @@
+								)),
+								None => {
+									let report_fut = net.report_peer(
+										peer.clone(),
+										UNKNOWN_PROTO_COST,
+									);
+
+									report_fut.await?;
+									None
+								}
+							};
+
+							if let Some(message) = message {
+								outgoing_messages.push(message);
+							}
+						}
+					}
+				}
+
+				let send_messages = ctx.send_messages(outgoing_messages);
+				if let Err(e) = send_messages.await {
+					log::warn!("Aborting - Failure to dispatch messages to overseer");
+					return Err(e)
+				}
+			},
+
+			Action::Abort => return Ok(()),
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use futures::channel::mpsc;
+	use futures::executor::{self, ThreadPool};
+
+	use std::sync::Arc;
+	use parking_lot::Mutex;
+	use assert_matches::assert_matches;
+
+	use polkadot_subsystem::messages::{StatementDistributionMessage, BitfieldDistributionMessage};
+
+	// The subsystem's view of the network - only supports a single call to `event_stream`.
+	#[derive(Clone)]
+	struct TestNetwork {
+		net_events: Arc<Mutex<Option<mpsc::UnboundedReceiver<NetworkEvent>>>>,
+		action_tx: mpsc::UnboundedSender<NetworkAction>,
+	}
+
+	// The test's view of the network. This receives updates from the subsystem in the form
+	// of `NetworkAction`s.
+	struct TestNetworkHandle {
+		action_rx: mpsc::UnboundedReceiver<NetworkAction>,
+		net_tx: mpsc::UnboundedSender<NetworkEvent>,
+	}
+
+	// a record of an action internal to the network.
+	#[derive(Debug)]
+	enum InternalActionRecord {
+		RegisterEventProducer(ProtocolId),
+		SendMessage(Vec<PeerId>, ProtocolId, Vec<u8>),
+		ReportPeer(PeerId, ReputationChange),
+		StartWork(Hash),
+		StopWork(Hash),
+
+		PeerConnected(PeerId, ObservedRole),
+		PeerDisconnected(PeerId),
+		PeerMessages(PeerId, Vec<WireMessage>),
+
+		Abort,
+	}
+
+	impl<'a> From<&'a Action> for InternalActionRecord {
+		fn from(action: &'a Action) -> Self {
+			match *action {
+				Action::RegisterEventProducer(protocol, _)
+					=> InternalActionRecord::RegisterEventProducer(protocol),
+				Action::SendMessage(ref peers, protocol, ref message)
+					=> InternalActionRecord::SendMessage(peers.clone(), protocol, message.clone()),
+				Action::ReportPeer(ref peer, rep)
+					=> InternalActionRecord::ReportPeer(peer.clone(), rep.clone()),
+
+				Action::StartWork(hash) => InternalActionRecord::StartWork(hash),
+				Action::StopWork(hash) => InternalActionRecord::StopWork(hash),
+
+				Action::PeerConnected(ref peer, ref role)
+					=> InternalActionRecord::PeerConnected(peer.clone(), role.clone()),
+				Action::PeerDisconnected(ref peer)
+					=> InternalActionRecord::PeerDisconnected(peer.clone()),
+				Action::PeerMessages(ref peer, ref messages)
+					=> InternalActionRecord::PeerMessages(peer.clone(), messages.clone()),
+
+				Action::Abort => InternalActionRecord::Abort,
+			}
+		}
+	}
+
+	fn new_test_network() -> (
+		TestNetwork,
+		TestNetworkHandle,
+	) {
+		let (net_tx, net_rx) = mpsc::unbounded();
```

Comment:
Using `mpsc::channel(0)` here instead might surface possible dead locks due to something depending on the channel to have a capacity. Just a thought.